applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_nodal_displacement_response_function.cpp

```diff
@@ -178,32 +185,43 @@ namespace Kratos
     {
         KRATOS_TRY;
 
-        const DoubleVariableType& r_traced_dof =
-            KratosComponents<DoubleVariableType>::Get(mTracedDofLabel);
+        const ArrayVariableType& r_traced_dof =
+            KratosComponents<ArrayVariableType>::Get(mTracedDofLabel);
+
+
+
+        double response_value = 0.0;
+        ModelPart& response_part = rModelPart.GetSubModelPart(mResponsePartName);
+        for(auto& node_i : response_part.Nodes()){
+            // project displacement vector in the traced direction. As mResponseDirection is a normalized vector
+            // the result of the inner product is alreay the displacement value in the traced direction.
+            response_value += inner_prod(mResponseDirection, node_i.FastGetSolutionStepValue(r_traced_dof, 0));
+        }
 
-        return rModelPart.GetNode(mpTracedNode->Id()).FastGetSolutionStepValue(r_traced_dof, 0);
+        return response_value;
 
         KRATOS_CATCH("");
     }
 
-    /// Find one element which is bounded by the traced node. The element is needed for assembling the adjoint load.
-    void AdjointNodalDisplacementResponseFunction::GetNeighboringElementPointer()
+    /// Find one element which is bounded by one of the traced nodes. The elements are needed for assembling the adjoint load.
+    void AdjointNodalDisplacementResponseFunction::GetNeighboringElementNodeMap()
     {
         KRATOS_TRY;
 
-        for (auto elem_it = mrModelPart.Elements().ptr_begin(); elem_it != mrModelPart.Elements().ptr_end(); ++elem_it)
-        {
-            const SizeType number_of_nodes = (*elem_it)->GetGeometry().PointsNumber();
-            for(IndexType i = 0; i < number_of_nodes; ++i)
-            {
-                if((*elem_it)->GetGeometry()[i].Id() == mpTracedNode->Id())
-                {
-                    mpNeighboringElement = (*elem_it);
-                    return;
+        ModelPart& response_part = mrModelPart.GetSubModelPart(mResponsePartName);
+        FindElementalNeighboursProcess neighbour_elements_finder(mrModelPart, 10, 10);
+        neighbour_elements_finder.Execute();
+
+        for(auto& node_i : response_part.Nodes()){
+            auto const& r_neighbours = node_i.GetValue(NEIGHBOUR_ELEMENTS);
+            // take the first element since only one neighbour element is required
+            const SizeType number_of_nodes = r_neighbours[0].GetGeometry().PointsNumber();
+            for(IndexType i = 0; i < number_of_nodes; ++i) {
+                if(r_neighbours[0].GetGeometry()[i].Id() == node_i.Id()) {
+                    mElementNodeMap[{r_neighbours[0].Id(), i}] = node_i.Id();
```

Comment:
Why do you need to store `i`?

Reply:
The idea was to have a unique key. But I can use the node id as key as well. Wasn't realizing this yesterday

Reply:
In that case i suggest a slightly different approach:
Store a `map<element_id, vector<node_ids>>`.
This has the advantage, that at the place you are using this information, you can directly access with the element_id. Because in the end you are interested in the what nodes you need to consider for a certain element.